Some additions to mongoose@4.0.0

nodkz opened the pull request:
### New
- Added [descriminator](http://mongoosejs.com/docs/discriminators.html) method definitions.
- Add `_indexes` property to `Mongoose$SchemaField` it keeps info about indexes over particular field.

### Fixes
- Changed `declare export type` to `declare type` due eslint error.

AndrewSouthpaw commented:
You'll need to explain the motivation for these fixes, so when other people using mongoose can understand your decisions. :)

nodkz commented:
@AndrewSouthpaw fixes are very tiny. Described them top message.

AndrewSouthpaw commented on definitions/npm/mongoose_v4.x.x/flow_v0.50.x-/mongoose_v4.x.x.js:
```diff
@@ -492,15 +495,15 @@ declare class Mongoose$Connection {
+  declare type MongooseDocument = Mongoose$Document;
+  declare type MongooseModel = typeof Mongoose$Document;
+  declare type MongooseSchema<Doc> = Mongoose$Schema<Doc>;
+  declare type MongooseSchemaField<Schema> = Mongoose$SchemaField<
```
Explain more about this decision. By removing these types from exports, this would potentially break consumers. Why is it necessary to stop exporting them?

nodkz commented:
Hm, very strange. Cause thwy are working in my project. I removed them because get errors from the pre commit checks.

So i can return then back if you ask.

AndrewSouthpaw commented:
Pre commit checks in your own project? You should just ignore the flow-typed library in eslint if that's what's happening.

nodkz commented:
Nope in pre-commig flow-typed repo. I'll send you tomorrow screenshot. From mobile now.

nodkz commented:
@AndrewSouthpaw sorry for delay had a heavy sprint at work.

I return back `export` statement and when commit them I receive eslint error on pre-commit hook:
<img width="479" alt="screen shot 2018-04-23 at 10 52 13" src="https://user-images.githubusercontent.com/1946920/39107564-fddb7fa0-46e4-11e8-95a4-79a8574c492f.png">

I removed `export` cause eslint checks did not allow me to made a commit.
So I made a commit with bypassing pre-hooks. 

Added after 10 minutes:
Looking on CI tests, eslint checks passed. But in dev machines `eslint` throws an error.

AndrewSouthpaw commented:
I was on vacation, sorry for the delay. 😄

I would suspect this is an issue with your eslint setup, since my eslinter never catches anything wrong with flow-typed. We can't approve changes based on your eslinting experience unless others report the same problem.

AndrewSouthpaw commented:
Ah, didn't actually check. Thanks! I'll keep an eye out for anyone else complaining about ESLint issues.
